runtime/src/main/java/com/dylibso/chicory/runtime/Memory.java

```diff
@@ -120,228 +116,191 @@ public void reinstantiate() {
         }
     }
 
-    public void copy(int dest, int src, int size) {
-        var data = new byte[size];
-        this.buffer.get(data, src, size);
-        try {
-            // TODO why can't i just write this array to the buffer without the loop?
-            for (var i = 0; i < size; i++) {
-                this.buffer.put(dest + i, data[i]);
-            }
-        } catch (IndexOutOfBoundsException e) {
-            throw new WASMRuntimeException("out of bounds memory access");
-        }
-    }
-
     public void initPassiveSegment(int segmentId, int dest, int offset, int size) {
         var segment = dataSegments[segmentId];
         if (!(segment instanceof PassiveDataSegment)) {
             throw new ChicoryException(
                     "data segment with id "
                             + " is not a passive segment and cannot be initialized at runtime");
         }
-        var data = segment.getData();
-        // TODO why doesn't this API work?
-        // this.buffer.put(target, segment.getData(), offset, size);
-        int j = dest;
-        try {
-            for (var i = offset; i < size; i++) {
-                this.buffer.put(j++, data[i]);
-            }
-        } catch (IndexOutOfBoundsException e) {
-            throw new WASMRuntimeException("out of bounds memory access");
-        }
+        write(dest, segment.getData(), offset, size);
     }
 
-    public String getString(int offset, int len) {
-        try {
-            var data = new byte[len];
-            for (int i = 0, j = offset; i < len; i++, j++) {
-                data[i] = this.buffer.get(j);
-            }
-
-            return new String(data);
-        } catch (IndexOutOfBoundsException e) {
-            throw new WASMRuntimeException("out of bounds memory access");
-        }
+    public void writeString(int offset, String data) {
+        write(offset, data.getBytes(StandardCharsets.UTF_8));
     }
 
-    public void put(int offset, String data) {
+    public String readString(int addr, int len) {
         try {
-            var bytes = data.getBytes(StandardCharsets.UTF_8);
-            for (int i = 0, j = offset; i < bytes.length; i++, j++) {
-                byte b = bytes[i];
-                this.buffer.put(j, b);
-            }
+            return new String(readBytes(addr, len), StandardCharsets.UTF_8);
         } catch (IndexOutOfBoundsException e) {
             throw new WASMRuntimeException("out of bounds memory access");
         }
     }
 
-    public void put(int offset, byte[] data) {
+    public void write(int addr, byte[] data) {
+        write(addr, data, 0, data.length);
+    }
+
+    public void write(int addr, byte[] data, int offset, int size) {
         try {
-            // System.out.println("mem-write@" + offset + " " + data);
```

Comment:
Thank you for fixing this! Couldn't find the right API for this.

Reply:
Yeah, `ByteBuffer` is tricky with that API inconsistency - I seem to recall running into it before, and it's one of the reasons I don't particularly like using `ByteBuffer` outside of I/O (though I think its use here is as good as we can do outside of using Panama's native memory API, which isn't even finalized until JDK 22).